svm/src/account_loader.rs

```diff
@@ -2227,4 +2371,271 @@ mod tests {
+
+        let program1_keypair = Keypair::new();
+        let program1 = program1_keypair.pubkey();
+        let (program1_size, _) = mk_account(program1, bpf_loader::id(), true);
+
+        let program2_keypair = Keypair::new();
+        let program2 = program2_keypair.pubkey();
+        let (program2_size, _) = mk_account(program2, bpf_loader_upgradeable::id(), true);
+
+        let programdata2_keypair = Keypair::new();
+        let programdata2 = programdata2_keypair.pubkey();
+        let (programdata2_size, _) = mk_account(programdata2, bpf_loader_upgradeable::id(), false);
+
+        let fee_payer_keypair = Keypair::new();
+        let fee_payer = fee_payer_keypair.pubkey();
+        let (fee_payer_size, fee_payer_account) =
+            mk_account(fee_payer, system_program::id(), false);
+
+        let account1_keypair = Keypair::new();
+        let account1 = account1_keypair.pubkey();
+        let (account1_size, _) = mk_account(account1, program1, false);
+
+        let account2_keypair = Keypair::new();
+        let account2 = account2_keypair.pubkey();
+        let (account2_size, _) = mk_account(account2, program2, false);
+
+        let test_data_size_with_cache = |instructions: Vec<_>, cache, expected_size| {
+            let transaction = SanitizedTransaction::from_transaction_for_tests(
+                Transaction::new_signed_with_payer(
+                    &instructions,
+                    Some(&fee_payer),
+                    &[&fee_payer_keypair],
+                    Hash::default(),
+                ),
+            );
+
+            let loaded_transaction_accounts = load_transaction_accounts(
+                &mock_bank,
+                &transaction,
+                LoadedTransactionAccount {
+                    account: fee_payer_account.clone(),
+                    loaded_size: fee_payer_size as usize,
+                    rent_collected: 0,
+                },
+                &ComputeBudgetLimits::default(),
+                &mut TransactionErrorMetrics::default(),
+                None,
+                &FeatureSet::default(),
+                &RentCollector::default(),
+                &cache,
+            )
+            .unwrap();
+
+            assert_eq!(
+                loaded_transaction_accounts.loaded_accounts_data_size,
+                expected_size
+            );
+        };
+
+        for account_meta in [AccountMeta::new, AccountMeta::new_readonly] {
+            let test_data_size = |instructions, expected_size| {
+                test_data_size_with_cache(
+                    instructions,
+                    ProgramCacheForTxBatch::default(),
+                    expected_size,
+                )
+            };
+
+            // one program plus loader
+            let ixns = vec![Instruction::new_with_bytes(program1, &[], vec![])];
+            test_data_size(ixns, program1_size + bpf_loader_size + fee_payer_size);
```

Comment:
is there any difference between the 2 loaders (except for size)?
should we also have the case "one program plus loader" for the upgradable loader program, i.e. program2?

Reply:
the loaders are functionally the same. the only thing unusual about the upgradeable loader is that its cached programs include programdata in their size